src/Fabulous/Memo.fs

```diff
@@ -0,0 +1,101 @@
+namespace Fabulous
+
+open Fabulous
+
+module Memo =
+
+    type internal MemoData =
+        {
+            /// Captures type of data that memoization depends on
+            KeyData: obj
+
+            // comparer that remembers KeyType internally
+            KeyComparer: obj -> obj -> bool
+
+            /// Lambda that users provide
+            CreateWidget: obj -> Widget
+
+            /// Captures type hash of data that memoization depends on
+            KeyTypeHash: int
+
+            /// Captures type hash of the marker memoized function produces
+            MarkerTypeHash: int
+        }
+
+    type Memoized<'t> = { phantom: 't }
+
+    let private MemoAttributeKey = AttributeDefinitionStore.getNextKey()
+    let internal MemoWidgetKey = WidgetDefinitionStore.getNextKey()
+
+    let inline private getMemoData (widget: Widget) : MemoData =
+        (widget.ScalarAttributes
+         |> Array.find(fun a -> a.Key = MemoAttributeKey))
+            .Value
+        :?> MemoData
+
+    let internal canReuseMemoizedViewNode prev next =
+        (getMemoData prev).MarkerTypeHash = (getMemoData next).MarkerTypeHash
+
+    let private compareAttributes (prev: MemoData, next: MemoData) : ScalarAttributeComparison =
+        match (prev.KeyTypeHash = next.KeyTypeHash, prev.MarkerTypeHash = next.MarkerTypeHash) with
+        | (true, true) ->
+            match next.KeyComparer next.KeyData prev.KeyData with
+            | true -> ScalarAttributeComparison.Identical
+            | false -> ScalarAttributeComparison.Different null
```

Comment:
Currently the diffvalue is not even passed into `Update`, the original intent was this use case:
1. Diffing an attribute can be expensive
2. Update requires information from the diffing two attributes
3. Thus, to avoid comparing attributes again you can pass value from `Compare` to `Update` like so `ScalarAttributeComparison.Different ( box "very expensive computation")`

But it didn't come up (yet?), so I'm wondering if we should just kill the ability to add a value to `Different` case

Reply:
Yes, I guess we should remove the value from `Different`.
My understanding was that `Compare` would also compute the diffing between 2 values, but with the separation of scalar/widget/widgetcolls, scalar attributes will always return the new value if different. No point in keeping this.